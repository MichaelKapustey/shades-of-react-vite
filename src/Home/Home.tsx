import Sheet from '@mui/joy/Sheet';
import Table from '@mui/joy/Table';
import Avatar from '@mui/joy/Avatar';
import Input from '@mui/joy/Input';
import Typography from '@mui/joy/Typography';

import request from 'graphql-request';
import { useQuery } from '@tanstack/react-query';
import { graphql } from '../gql/gql';
import { useState } from 'react';
import { Pagination } from '../components/Pagination';
import { Link } from 'react-router-dom';
import { Page } from '../components/Page';

const allCharacters = graphql(/* GraphQL */ `
query getCharacters($searchValue: String!, $page: Int!){
  characters(page: $page, filter:{name: $searchValue }){
    info{
      count,
      pages
  },
    results{
      id,
      name,
      status,
      species,
      image
    }
  }
}
`);

export const Home = () => {
  const [search, setSearch] = useState('');
  const [page, setPage] = useState(1);
  const { data } = useQuery({
    queryKey: ['characters', search, page],
    queryFn: async ({ queryKey }) =>
      request(
        'https://rickandmortyapi.com/graphql',
        allCharacters,
        { searchValue: queryKey[1] as string, page: queryKey[2] as number }
      ).catch(err => {
        console.error(err);
        return;
      }),
    placeholderData: (prev) => prev,
  })

  return <Page>
    <Typography level='h1' variant='solid' color='primary' textAlign="center">
      Characters
    </Typography>
    <Input
      placeholder='Type to search'
      value={search}
      onChange={(e) => {
        setSearch(e.target.value);
        setPage(1);
      }}>
    </Input>
    <Sheet sx={{ overflowY: 'auto', mb: 'auto' }}>
      <Table>
        <thead>
          <tr>
            <th>Avatar</th>
            <th>Name</th>
            <th>Species</th>
            <th>Status</th>
          </tr>
        </thead>
        <tbody data-testid="characters-table-body">
          {data?.characters?.results?.map((row) => (
            <tr key={row?.id}>
              <td><Avatar src={row?.image ?? ''} alt={row?.name ?? ''} /></td>
              <td><Typography
                variant='plain'
                level='body-sm'
                sx={{ textDecoration: 'none' }}
                component={Link}
                to={`/${row?.id}`}>
                {row?.name}
              </Typography>
              </td>
              <td><Typography variant='plain' level='body-sm'>{row?.species}</Typography></td>
              <td><Typography variant='plain' level='body-sm'>{row?.status}</Typography></td>
            </tr>
          ))}
        </tbody>
      </Table>

    </Sheet >
    <Pagination current={page} total={data?.characters?.info?.pages || 0} onPageChange={(p) => setPage(p)} />
  </Page >

}