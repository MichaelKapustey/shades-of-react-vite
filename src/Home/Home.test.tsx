import { render } from '@testing-library/react';
import { Home } from './Home';
import { vi } from 'vitest';
import React from 'react';

// #region Mocks

vi.mock('@tanstack/react-query', () => ({
    useQuery: () => ({
        data: {
            characters: {
                info: {
                    count: 1,
                    pages: 1,
                },
                results: [{
                    name: 'Char 1',
                    image: '/1.png',
                    status: 'Alive',
                    species: 'Human'
                }, {
                    name: 'Char 2',
                    image: '/2.png',
                    status: 'Dead',
                    species: 'Alien'
                }]
            },
        },
    })
}))

vi.mock('react-router-dom', () => ({
    Link: ({ children, ...rest }: React.PropsWithChildren) => <a {...rest}>{children}</a>
}))

vi.mock('@mui/joy/Sheet', () => ({
    default: ({ children, ...rest }: React.PropsWithChildren) => <div {...rest}>{children}</div>
}))

vi.mock('@mui/joy/Table', () => ({
    default: ({ children, ...rest }: React.PropsWithChildren) => <table {...rest}>{children}</table>
}))

vi.mock('@mui/joy/Avatar', () => ({
    default: ({ children, ...rest }: React.PropsWithChildren) => <img {...rest}>{children}</img>
}))

vi.mock('@mui/joy/Typography', () => ({
    default: ({ children, ...rest }: React.PropsWithChildren) => <p {...rest}>{children}</p>
}))

vi.mock('@mui/joy/Input', () => ({
    default: ({ children, ...rest }: React.PropsWithChildren) => <input {...rest}>{children}</input>
}))

vi.mock('../components/TinyCard', () => ({
    TinyCard: ({ children, ...rest }: React.PropsWithChildren) => <div {...rest}>{children}</div>
}))

vi.mock('../components/Page', () => ({
    Page: ({ children, ...rest }: React.PropsWithChildren) => <div {...rest}>{children}</div>
}))

vi.mock('../components/Pagination', () => ({
    Pagination: ({ children, ...rest }: React.PropsWithChildren) => <div {...rest}>{children}</div>
}))

//#endregion

test('Home renders correctly and matches snapshot', () => {
    const { asFragment } = render(<Home />);
    expect(asFragment()).toMatchSnapshot();
});

