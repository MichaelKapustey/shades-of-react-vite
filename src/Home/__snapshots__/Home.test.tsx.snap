// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Home renders correctly and matches snapshot 1`] = `
<DocumentFragment>
  <div>
    <p
      color="primary"
      level="h1"
      textalign="center"
      variant="solid"
    >
      Characters
    </p>
    <input
      placeholder="Type to search"
      value=""
    />
    <div
      sx="[object Object]"
    >
      <table>
        <thead>
          <tr>
            <th>
              Avatar
            </th>
            <th>
              Name
            </th>
            <th>
              Species
            </th>
            <th>
              Status
            </th>
          </tr>
        </thead>
        <tbody
          data-testid="characters-table-body"
        >
          <tr>
            <td>
              <img
                alt="Char 1"
                src="/1.png"
              />
            </td>
            <td>
              <p
                level="body-sm"
                sx="[object Object]"
                to="/undefined"
                variant="plain"
              >
                Char 1
              </p>
            </td>
            <td>
              <p
                level="body-sm"
                variant="plain"
              >
                Human
              </p>
            </td>
            <td>
              <p
                level="body-sm"
                variant="plain"
              >
                Alive
              </p>
            </td>
          </tr>
          <tr>
            <td>
              <img
                alt="Char 2"
                src="/2.png"
              />
            </td>
            <td>
              <p
                level="body-sm"
                sx="[object Object]"
                to="/undefined"
                variant="plain"
              >
                Char 2
              </p>
            </td>
            <td>
              <p
                level="body-sm"
                variant="plain"
              >
                Alien
              </p>
            </td>
            <td>
              <p
                level="body-sm"
                variant="plain"
              >
                Dead
              </p>
            </td>
          </tr>
        </tbody>
      </table>
    </div>
    <div
      current="1"
      total="1"
    />
  </div>
</DocumentFragment>
`;
